File: android/ksdialogs/src/main/kotlin/jp/kamusoft/ksdialogs/ToastViewRegistry.kt
package jp.kamusoft.ksdialogs

import android.content.Context
import android.view.View
import kotlin.reflect.KClass

/**
 * カスタム Toast の ViewModel 型をキーに View factory と ViewModel factory を引くレジストリ。
 *
 * Dialog のレジストリ ([DialogViewRegistry]) とも Loading のレジストリ ([LoadingViewRegistry]) とも
 * 独立しており、同じ ViewModel 型をそれぞれへ別々の View で登録できる。
 * 片方の再登録は他方に影響しない。登録・解決は任意のスレッドから行える。
 *
 * 1つの ViewModel 型のエントリは View factory と ViewModel factory の2スロットからなり、
 * 再登録はスロット単位の後勝ちで、もう片方のスロットは保持される。
 */
public class ToastViewRegistry {
    private val lock = Any()
    private val entries = mutableMapOf<KClass<*>, ToastRegistryEntry>()

    /**
     * ViewModel 型に対する従来 View 系の factory を登録する。
     *
     * 同じ型の View factory への再登録は後勝ちで置き換え、ViewModel factory は保持される。
     * factory は表示のたびに呼ばれ、View を毎回新規に生成する。
     * factory のレシーバは、View の生成に使う提示先画面の [Context]。
     *
     * @param viewModelClass 登録キーになる ViewModel のクラス参照
     * @param factory 中身の View を生成する関数
     */
    public fun <VM : ToastViewModel> register(
        viewModelClass: KClass<VM>,
        factory: Context.(VM) -> View,
    ) {
        register(viewModelClass, erasedToastViewFactory(factory))
    }

    /**
     * ViewModel 型に対する ViewModel factory を登録する。
     *
     * 型指定の表示 (クラス参照を渡す [KsToast.show]) はこの factory で ViewModel を作る。
     * 同じ型の ViewModel factory への再登録は後勝ちで置き換え、View factory は保持される。
     * factory は表示のたびに呼ばれ、UI スレッドで実行される。
     *
     * @param viewModelClass 登録キーになる ViewModel のクラス参照
     * @param factory ViewModel を生成する関数
     */
    public fun <VM : ToastViewModel> registerViewModel(
        viewModelClass: KClass<VM>,
        factory: () -> VM,
    ) {
        viewModelClass.requireReferenceTypeViewModel()
        val erasedFactory = ToastViewModelFactory { factory() }
        updateEntry(viewModelClass) { it.copy(viewModelFactory = erasedFactory) }
    }

    /** 型消去された View factory を登録する。宣言的 UI 系の登録面もこの口を通る。 */
    internal fun register(viewModelClass: KClass<*>, factory: ToastViewFactory) {
        viewModelClass.requireReferenceTypeViewModel()
        updateEntry(viewModelClass) { it.copy(viewFactory = factory) }
    }

    /** エントリの片方のスロットだけを置き換える。未登録の型なら空のエントリから作る。 */
    private fun updateEntry(
        viewModelClass: KClass<*>,
        update: (ToastRegistryEntry) -> ToastRegistryEntry,
    ) {
        synchronized(lock) {
            entries[viewModelClass] = update(entries[viewModelClass] ?: ToastRegistryEntry())
        }
    }

    /**
     * キーに対応するエントリを返す。どちらのスロットも未登録なら null。
     * 返すのは呼び出し時点のスナップショットで、以後の再登録には影響されない。
     */
    internal fun entry(viewModelClass: KClass<*>): ToastRegistryEntry? =
        synchronized(lock) { entries[viewModelClass] }

    /**
     * キーに対応する View factory を返す。未登録なら null。
     * 返すのは呼び出し時点のスナップショットで、以後の再登録には影響されない。
     */
    internal fun factory(viewModelClass: KClass<*>): ToastViewFactory? =
        entry(viewModelClass)?.viewFactory

    public companion object {
        /** 全入口が共有する既定のレジストリ。 */
        public val shared: ToastViewRegistry = ToastViewRegistry()
    }
}

File: maui/KsDialogs.Maui/Registry/ToastViewRegistry.cs
using System;
using System.Collections.Generic;
using System.Threading;
using Microsoft.Maui.Controls;

namespace KsDialogs;

/// <summary>
/// カスタム Toast の ViewModel 型をキーに MAUI View factory と ViewModel factory を引くレジストリ。
/// </summary>
/// <remarks>
/// Dialog のレジストリ (<see cref="DialogViewRegistry"/>) とも Loading のレジストリ
/// (<see cref="LoadingViewRegistry"/>) とも独立しており、同じ ViewModel 型をそれぞれへ別々の View で
/// 登録できる。片方の再登録は他方に影響しない。
/// 既定 singleton エントリ (<see cref="Toast.Instance"/>) と DI 注入で使うインスタンスは
/// <see cref="Shared"/> を共有するため、どちらの入口から登録しても同じ紐付けが引ける。
/// 登録・解決は任意のスレッドから行える。
/// <para>
/// このレジストリは MAUI 形態の層に属し、Native ライブラリ側のレジストリとは別物である
/// (maui/ADR-0001)。
/// </para>
/// <para>
/// 1 つの ViewModel 型のエントリは View factory と ViewModel factory の 2 スロットからなり、
/// 再登録はスロット単位の後勝ちで、もう片方のスロットは保持される。
/// </para>
/// </remarks>
public sealed class ToastViewRegistry
{
    /// <summary>全入口が共有する既定のレジストリ。</summary>
    public static ToastViewRegistry Shared { get; } = new();

    private readonly Lock _gate = new();
    private readonly Dictionary<Type, ToastRegistryEntry> _entries = [];

    /// <summary>
    /// ViewModel 型に対する MAUI View factory を登録する。同じスロットへの再登録は後勝ちで置き換える。
    /// </summary>
    /// <remarks>factory は表示のたびに呼ばれ、View を毎回新規に生成する (core/ADR-0005)。</remarks>
    /// <typeparam name="TViewModel">登録キーになる ViewModel の型。</typeparam>
    /// <param name="factory">中身の MAUI View を生成する関数。</param>
    public void Register<TViewModel>(Func<TViewModel, View> factory)
        where TViewModel : class, IToastViewModel
    {
        ArgumentNullException.ThrowIfNull(factory);

        StoreViewFactory(typeof(TViewModel), ToastViewFactories.Erase(factory));
    }

    /// <summary>
    /// ViewModel 型に対する ViewModel factory を登録する。
    /// </summary>
    /// <remarks>
    /// 型指定 show (<c>Show&lt;TViewModel&gt;()</c> の形) はこの factory で ViewModel を作る。
    /// factory は表示のたびに呼ばれ、View factory と同じく表示が受理されたあとの UI スレッドで
    /// 実行される。View factory とはスロットが別なので、どちらを登録し直しても他方は保持される。
    /// </remarks>
    /// <typeparam name="TViewModel">登録キーになる ViewModel の型。</typeparam>
    /// <param name="factory">ViewModel を生成する関数。</param>
    public void RegisterViewModel<TViewModel>(Func<TViewModel> factory)
        where TViewModel : class, IToastViewModel
    {
        ArgumentNullException.ThrowIfNull(factory);

        StoreViewModelFactory(typeof(TViewModel), () => factory());
    }

    /// <summary>型消去した View factory を該当スロットへ入れる。</summary>
    /// <param name="viewModelType">登録キーになる ViewModel の型。</param>
    /// <param name="factory">型消去した View factory。</param>
    internal void StoreViewFactory(Type viewModelType, ToastViewFactory factory) =>
        UpdateEntry(viewModelType, entry => entry with { ViewFactory = factory });

    /// <summary>型消去した ViewModel factory を該当スロットへ入れる。</summary>
    /// <param name="viewModelType">登録キーになる ViewModel の型。</param>
    /// <param name="factory">型消去した ViewModel factory。</param>
    internal void StoreViewModelFactory(Type viewModelType, ToastViewModelFactory factory) =>
        UpdateEntry(viewModelType, entry => entry with { ViewModelFactory = factory });

    /// <summary>
    /// キーに対応するエントリを返す。どちらのスロットも未登録なら <see langword="null"/>。
    /// </summary>
    /// <remarks>返すのは呼び出し時点のスナップショットで、以後の再登録には影響されない。</remarks>
    /// <param name="viewModelType">解決したい ViewModel の型。</param>
    /// <returns>登録済みのエントリ。未登録なら <see langword="null"/>。</returns>
    internal ToastRegistryEntry? Entry(Type viewModelType)
    {
        lock (_gate)
        {
            return _entries.TryGetValue(viewModelType, out ToastRegistryEntry? entry) ? entry : null;
        }
    }

    /// <summary>キーに対応する View factory を返す。未登録なら <see langword="null"/>。</summary>
    /// <param name="viewModelType">解決したい ViewModel の型。</param>
    /// <returns>登録済みの View factory。未登録なら <see langword="null"/>。</returns>
    internal ToastViewFactory? Factory(Type viewModelType) => Entry(viewModelType)?.ViewFactory;

    /// <summary>エントリの片方のスロットだけを置き換える。未登録の型なら空のエントリから作る。</summary>
    private void UpdateEntry(Type viewModelType, Func<ToastRegistryEntry, ToastRegistryEntry> update)
    {
        lock (_gate)
        {
            _entries[viewModelType] =
                update(_entries.TryGetValue(viewModelType, out ToastRegistryEntry? entry) ? entry : new());
        }
    }
}
